File: packages/contracts/python/demodirector_contracts/evidence.py
from datetime import datetime
from typing import Literal

from pydantic import Field, HttpUrl, model_validator

from demodirector_contracts.models import ContractModel, NonEmptyString, ResearchSource


class NarrationEvidence(ContractModel):
    id: NonEmptyString
    scene_id: NonEmptyString
    text: NonEmptyString
    status: Literal["source_quote", "source_linked", "user_assertion", "unverified"]
    source_ids: list[NonEmptyString]
    explanation: NonEmptyString


class ContributionSource(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    title: NonEmptyString
    url: HttpUrl | None = None
    domain: NonEmptyString
    origin: Literal["project_brief", "website_inspection", "parallel_search"]
    retrieval_state: Literal["saved", "partial"]
    retrieved_at: datetime | None = None
    excerpt: str = Field(min_length=1, max_length=500)


class ContributionScene(ContractModel):
    id: NonEmptyString
    title: NonEmptyString
    order: int = Field(ge=0)


class SourceContribution(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    source_id: NonEmptyString
    kind: Literal[
        "brief_context", "website_structure", "research_context", "factual_narration"
    ]
    scene_ids: list[NonEmptyString] = Field(default_factory=list)
    narration_statement_ids: list[NonEmptyString] = Field(default_factory=list)
    usage_state: Literal["used", "unused"]
    label: NonEmptyString


class SourceContributionMap(ContractModel):
    project_id: NonEmptyString
    storyboard_version: int = Field(gt=0)
    fingerprint: str = Field(pattern=r"^[a-f0-9]{64}$")
    status: Literal["ready", "stale", "approval_required"]
    partial_evidence: bool
    sources: list[ContributionSource]
    scenes: list[ContributionScene]
    narration_statements: list[NarrationEvidence]
    contributions: list[SourceContribution]

    @model_validator(mode="after")
    def references_are_project_owned_and_supported(self) -> "SourceContributionMap":
        source_by_id = {source.id: source for source in self.sources}
        scene_ids = {scene.id for scene in self.scenes}
        statement_by_id = {statement.id: statement for statement in self.narration_statements}
        if len(source_by_id) != len(self.sources) or len(scene_ids) != len(self.scenes):
            raise ValueError("Contribution source and scene IDs must be unique")
        if len(statement_by_id) != len(self.narration_statements):
            raise ValueError("Narration statement IDs must be unique")
        if any(source.project_id != self.project_id for source in self.sources):
            raise ValueError("Contribution sources must belong to the project")
        if self.status != "ready" and (self.contributions or self.narration_statements):
            raise ValueError("Unapproved contribution maps must not retain usage references")
        for statement in self.narration_statements:
            if (
                statement.scene_id not in scene_ids
                or set(statement.source_ids) - source_by_id.keys()
            ):
                raise ValueError("Narration statements must reference known project evidence")
        contribution_ids: set[str] = set()
        for contribution in self.contributions:
            if contribution.id in contribution_ids:
                raise ValueError("Contribution IDs must be unique")
            contribution_ids.add(contribution.id)
            if contribution.project_id != self.project_id:
                raise ValueError("Contributions must belong to the project")
            if contribution.source_id not in source_by_id:
                raise ValueError("Contributions must reference known project evidence")
            if set(contribution.scene_ids) - scene_ids:
                raise ValueError("Contributions must reference known storyboard scenes")
            if set(contribution.narration_statement_ids) - statement_by_id.keys():
                raise ValueError("Contributions must reference known narration statements")
            has_usage = bool(contribution.scene_ids or contribution.narration_statement_ids)
            if (contribution.usage_state == "used") != has_usage:
                raise ValueError("Contribution usage state must match its references")
            if contribution.kind == "website_structure" and contribution.narration_statement_ids:
                raise ValueError("Website structure is a capture input, not a factual claim")
            if contribution.kind == "factual_narration":
                if not contribution.narration_statement_ids:
                    raise ValueError("Factual narration contributions require statements")
                for statement_id in contribution.narration_statement_ids:
                    statement = statement_by_id[statement_id]
                    if (
                        statement.status not in {"source_quote", "source_linked"}
                        or contribution.source_id not in statement.source_ids
                        or statement.scene_id not in contribution.scene_ids
                    ):
                        raise ValueError("Factual narration must be supported by saved evidence")
        return self


class StoryboardEvidence(ContractModel):
    project_id: NonEmptyString
    storyboard_version: int = Field(gt=0)
    fingerprint: str = Field(pattern=r"^[a-f0-9]{64}$")
    claims: list[NarrationEvidence]
    sources: list[ResearchSource]
    unverified_count: int = Field(ge=0)
    approved: bool
    approval_required: bool = True

File: packages/contracts/src/evidence.ts
import { z } from "zod";
import { researchSourceSchema } from "./schemas";
const text = z.string().min(1);
export const narrationEvidenceSchema = z.object({
  id: text, scene_id: text, text,
  status: z.enum(["source_quote", "source_linked", "user_assertion", "unverified"]),
  source_ids: z.array(text), explanation: text,
}).strict();
export const storyboardEvidenceSchema = z.object({
  project_id: z.string().min(1), storyboard_version: z.number().int().positive(),
  fingerprint: z.string().regex(/^[a-f0-9]{64}$/),
  claims: z.array(narrationEvidenceSchema),
  sources: z.array(researchSourceSchema), unverified_count: z.number().int().nonnegative(),
  approved: z.boolean(), approval_required: z.boolean(),
}).strict().superRefine((value, ctx) => {
  const known = new Set(value.sources.map((s) => s.id));
  if (value.sources.some((s) => s.project_id !== value.project_id) || value.claims.some((c) => c.source_ids.some((id) => !known.has(id))) || value.unverified_count !== value.claims.filter((c) => c.status === "unverified" || c.status === "source_linked").length) {
    ctx.addIssue({ code: "custom", message: "Evidence references or unverified count are invalid" });
  }
});
export type StoryboardEvidence = z.infer<typeof storyboardEvidenceSchema>;

export const contributionSourceSchema = z.object({
  id: text,
  project_id: text,
  title: text,
  url: z.string().url().nullable(),
  domain: text,
  origin: z.enum(["project_brief", "website_inspection", "parallel_search"]),
  retrieval_state: z.enum(["saved", "partial"]),
  retrieved_at: z.string().datetime().nullable(),
  excerpt: text.max(500),
}).strict();

export const contributionSceneSchema = z.object({
  id: text,
  title: text,
  order: z.number().int().nonnegative(),
}).strict();

export const sourceContributionSchema = z.object({
  id: text,
  project_id: text,
  source_id: text,
  kind: z.enum(["brief_context", "website_structure", "research_context", "factual_narration"]),
  scene_ids: z.array(text),
  narration_statement_ids: z.array(text),
  usage_state: z.enum(["used", "unused"]),
  label: text,
}).strict();

export const sourceContributionMapSchema = z.object({
  project_id: text,
  storyboard_version: z.number().int().positive(),
  fingerprint: z.string().regex(/^[a-f0-9]{64}$/),
  status: z.enum(["ready", "stale", "approval_required"]),
  partial_evidence: z.boolean(),
  sources: z.array(contributionSourceSchema),
  scenes: z.array(contributionSceneSchema),
  narration_statements: z.array(narrationEvidenceSchema),
  contributions: z.array(sourceContributionSchema),
}).strict().superRefine((value, ctx) => {
  const fail = (message: string) => ctx.addIssue({ code: "custom", message });
  const sourceIds = new Set(value.sources.map((source) => source.id));
  const sceneIds = new Set(value.scenes.map((scene) => scene.id));
  const statements = new Map(value.narration_statements.map((statement) => [statement.id, statement]));
  if (sourceIds.size !== value.sources.length || sceneIds.size !== value.scenes.length || statements.size !== value.narration_statements.length) fail("Contribution IDs must be unique");
  if (value.sources.some((source) => source.project_id !== value.project_id)) fail("Contribution sources must belong to the project");
  if (value.status !== "ready" && (value.contributions.length > 0 || value.narration_statements.length > 0)) fail("Stale maps must discard usage references");
  if (value.narration_statements.some((statement) => !sceneIds.has(statement.scene_id) || statement.source_ids.some((id) => !sourceIds.has(id)))) fail("Narration statements must use known evidence");
  const contributionIds = new Set<string>();
  for (const contribution of value.contributions) {
    if (contributionIds.has(contribution.id)) fail("Contribution IDs must be unique");
    contributionIds.add(contribution.id);
    if (contribution.project_id !== value.project_id || !sourceIds.has(contribution.source_id)) fail("Contribution must use project evidence");
    if (contribution.scene_ids.some((id) => !sceneIds.has(id)) || contribution.narration_statement_ids.some((id) => !statements.has(id))) fail("Contribution references are unknown");
    const used = contribution.scene_ids.length > 0 || contribution.narration_statement_ids.length > 0;
    if ((contribution.usage_state === "used") !== used) fail("Contribution usage state is inconsistent");
    if (contribution.kind === "website_structure" && contribution.narration_statement_ids.length > 0) fail("Website structure cannot support narration claims");
    if (contribution.kind === "factual_narration") {
      if (contribution.narration_statement_ids.length === 0) fail("Factual narration requires supported statements");
      for (const id of contribution.narration_statement_ids) {
        const statement = statements.get(id);
        if (!statement || !["source_quote", "source_linked"].includes(statement.status) || !statement.source_ids.includes(contribution.source_id) || !contribution.scene_ids.includes(statement.scene_id)) fail("Factual narration is not supported by saved evidence");
      }
    }
  }
});
export type SourceContributionMap = z.infer<typeof sourceContributionMapSchema>;
